File: src/Entity/Reponse.php
<?php

namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;
use App\Entity\Reclamation;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity]
#[ORM\Table(name: 'reponse')]
class Reponse
{
    #[ORM\Id]
    #[ORM\GeneratedValue(strategy: 'AUTO')]
    #[ORM\Column(name: 'id_reponse', type: 'integer')]
    private int $id_reponse = 0;

    #[ORM\ManyToOne(targetEntity: Reclamation::class, inversedBy: "reponses")]
   #[ORM\JoinColumn(name: 'reclamation_id', referencedColumnName: 'id_reclamation', nullable: false, onDelete: 'CASCADE')]
    #[Assert\NotNull(message: 'La réclamation associée est obligatoire')]
    private ?Reclamation $reclamation = null;

    #[ORM\Column(type: "text")]
    #[Assert\NotBlank(message: 'Le contenu de la réponse ne peut pas être vide')]
    #[Assert\Length(
        min: 5,
        max: 5000,
        minMessage: 'Le contenu doit contenir au moins {{ limit }} caractères',
        maxMessage: 'Le contenu ne doit pas dépasser {{ limit }} caractères'
    )]
    private string $contenu_reponse = '';

    #[ORM\Column(type: "datetime")]
    #[Assert\NotNull(message: 'La date de la réponse est obligatoire')]
    private \DateTimeInterface $date_reponse;

    #[ORM\Column(type: "integer", nullable: true)]
    #[Assert\NotNull(message: "L'ID utilisateur est obligatoire")]
    #[Assert\GreaterThan(0, message: "L'ID utilisateur doit être positif")]
    private ?int $id_utilisateur = null;

    #[ORM\Column(type: "string", length: 255, nullable: true)]
    private ?string $nom_utilisateur = null;

    #[ORM\Column(type: "float")]
    #[Assert\Range(
        min: 0,
        max: 5,
        notInRangeMessage: 'La note doit être entre {{ min }} et {{ max }}'
    )]
    private float $rate_reponse = 0;

    #[ORM\Column(type: "integer")]
    #[Assert\GreaterThanOrEqual(0, message: 'La somme des notes doit être positive ou zéro')]
    private int $rate_sum = 0;

    #[ORM\Column(type: "integer")]
    #[Assert\GreaterThanOrEqual(0, message: 'Le nombre de notes doit être positif ou zéro')]
    private int $rate_count = 0;

    public function __construct()
    {
        $this->date_reponse = new \DateTime();
        $this->id_utilisateur = null;
    }

    // ── Getters / Setters snake_case (compatibilité legacy) ───────────────────

    public function getId_reponse(): int
    {
        return $this->id_reponse;
    }

    public function getReclamation(): ?Reclamation
    {
        return $this->reclamation;
    }

    public function setReclamation(?Reclamation $reclamation): self
    {
        $this->reclamation = $reclamation;
        return $this;
    }

    public function getContenu_reponse(): string
    {
        return $this->contenu_reponse;
    }

    public function setContenu_reponse(string $value): self
    {
        $this->contenu_reponse = $value;
        return $this;
    }

    public function getDate_reponse(): \DateTimeInterface
    {
        return $this->date_reponse;
    }

    // protected — les timestamps ne doivent pas être manipulés manuellement
    protected function setDate_reponse(\DateTimeInterface $value): self
    {
        $this->date_reponse = $value;
        return $this;
    }

    public function getId_utilisateur(): ?int
    {
        return $this->id_utilisateur;
    }

    public function setId_utilisateur(?int $value): self
    {
        $this->id_utilisateur = $value;
        return $this;
    }

    public function getNom_utilisateur(): ?string
    {
        return $this->nom_utilisateur;
    }

    public function setNom_utilisateur(?string $value): self
    {
        $this->nom_utilisateur = $value;
        return $this;
    }

    public function getRate_reponse(): float
    {
        return $this->rate_reponse;
    }

    public function setRate_reponse(float $value): self
    {
        $this->rate_reponse = $value;
        return $this;
    }

    public function getRate_sum(): int
    {
        return $this->rate_sum;
    }

    public function setRate_sum(int $value): self
    {
        $this->rate_sum = $value;
        return $this;
    }

    public function getRate_count(): int
    {
        return $this->rate_count;
    }

    public function setRate_count(int $value): self
    {
        $this->rate_count = $value;
        return $this;
    }

    // ── Getters / Setters camelCase (PropertyAccess / EntityType / Symfony forms) ──

    public function getIdReponse(): int
    {
        return $this->id_reponse;
    }

    public function getContenuReponse(): string
    {
        return $this->contenu_reponse;
    }

    public function setContenuReponse(string $value): self
    {
        return $this->setContenu_reponse($value);
    }

    public function getDateReponse(): \DateTimeInterface
    {
        return $this->date_reponse;
    }

    // protected — timestamp géré automatiquement via le constructeur
    protected function setDateReponse(\DateTimeInterface $value): self
    {
        return $this->setDate_reponse($value);
    }

    public function getIdUtilisateur(): ?int
    {
        return $this->id_utilisateur;
    }

    public function setIdUtilisateur(?int $value): self
    {
        return $this->setId_utilisateur($value);
    }

    public function getNomUtilisateur(): ?string
    {
        return $this->nom_utilisateur;
    }

    public function setNomUtilisateur(?string $value): self
    {
        return $this->setNom_utilisateur($value);
    }

    public function getRateReponse(): float
    {
        return $this->rate_reponse;
    }

    public function setRateReponse(float $value): self
    {
        return $this->setRate_reponse($value);
    }

    public function getRateSum(): int
    {
        return $this->rate_sum;
    }

    public function setRateSum(int $value): self
    {
        return $this->setRate_sum($value);
    }

    public function getRateCount(): int
    {
        return $this->rate_count;
    }

    public function setRateCount(int $value): self
    {
        return $this->setRate_count($value);
    }
}

File: src/main/java/tn/esprit/entities/Reponse.java
package tn.esprit.entities;

import java.sql.Timestamp;

public class Reponse {

    private int idReponse;           // id_reponse        AUTO_INCREMENT
    private String contenuReponse;   // contenu_reponse   longtext
    private Timestamp dateReponse;   // date_reponse      datetime   ← ⚠️ (plus timestamp)
    private int idUtilisateur;       // id_utilisateur    int(11)    NULL
    private String nomUtilisateur;   // nom_utilisateur   varchar(255) NULL ← ⚠️ colonne 5 (avant rate)
    private double rateReponse;      // rate_reponse      double     ← ⚠️ double
    private int rateSum;             // rate_sum          int(11)
    private int rateCount;           // rate_count        int(11)
    private int reclamationId;       // reclamation_id    int(11)    ← ⚠️ renommé

    // ─── Constructeur complet (avec ID) ──────────────────────────────────────
    public Reponse(int idReponse, String contenuReponse, Timestamp dateReponse,
                   int idUtilisateur, String nomUtilisateur,
                   double rateReponse, int rateSum, int rateCount,
                   int reclamationId) {
        this.idReponse      = idReponse;
        this.contenuReponse = contenuReponse;
        this.dateReponse    = dateReponse;
        this.idUtilisateur  = idUtilisateur;
        this.nomUtilisateur = nomUtilisateur;
        this.rateReponse    = rateReponse;
        this.rateSum        = rateSum;
        this.rateCount      = rateCount;
        this.reclamationId  = reclamationId;
    }

    // ─── Constructeur sans ID (pour insertion) ────────────────────────────────
    public Reponse(String contenuReponse, Timestamp dateReponse,
                   int idUtilisateur, String nomUtilisateur,
                   double rateReponse, int reclamationId) {
        this.contenuReponse = contenuReponse;
        this.dateReponse    = dateReponse;
        this.idUtilisateur  = idUtilisateur;
        this.nomUtilisateur = nomUtilisateur;
        this.rateReponse    = rateReponse;
        this.reclamationId  = reclamationId;
    }

    // ─── Constructeur sans ID ni nomUtilisateur ───────────────────────────────
    public Reponse(String contenuReponse, Timestamp dateReponse,
                   int idUtilisateur, double rateReponse, int reclamationId) {
        this(contenuReponse, dateReponse, idUtilisateur, null, rateReponse, reclamationId);
    }

    // ─── Constructeur vide ────────────────────────────────────────────────────
    public Reponse() {}

    // ─── Getters & Setters ────────────────────────────────────────────────────

    public int getIdReponse() { return idReponse; }
    public void setIdReponse(int idReponse) { this.idReponse = idReponse; }

    public String getContenuReponse() { return contenuReponse; }
    public void setContenuReponse(String contenuReponse) { this.contenuReponse = contenuReponse; }

    public Timestamp getDateReponse() { return dateReponse; }
    public void setDateReponse(Timestamp dateReponse) { this.dateReponse = dateReponse; }

    public int getIdUtilisateur() { return idUtilisateur; }
    public void setIdUtilisateur(int idUtilisateur) { this.idUtilisateur = idUtilisateur; }

    /** Colonne {@code nom_utilisateur} en base. */
    public String getNomUtilisateur() { return nomUtilisateur; }
    public void setNomUtilisateur(String nomUtilisateur) { this.nomUtilisateur = nomUtilisateur; }

    /** Retourne la moyenne calculée si des votes existent, sinon la valeur stockée. */
    public double getRateReponse() {
        if (rateCount > 0) return (double) rateSum / rateCount;
        return rateReponse;
    }
    public void setRateReponse(double rateReponse) { this.rateReponse = rateReponse; }

    public int getRateSum() { return rateSum; }
    public void setRateSum(int rateSum) { this.rateSum = rateSum; }

    public int getRateCount() { return rateCount; }
    public void setRateCount(int rateCount) { this.rateCount = rateCount; }

    /** Colonne {@code reclamation_id} en base. */
    public int getReclamationId() { return reclamationId; }
    public void setReclamationId(int reclamationId) { this.reclamationId = reclamationId; }

    @Override
    public String toString() {
        return "Reponse{" +
                "idReponse=" + idReponse +
                ", contenuReponse='" + contenuReponse + '\'' +
                ", dateReponse=" + dateReponse +
                ", idUtilisateur=" + idUtilisateur +
                ", nomUtilisateur='" + nomUtilisateur + '\'' +
                ", reclamationId=" + reclamationId +
                ", rateReponse=" + getRateReponse() +
                '}';
    }
}